src/lib/remark-plugins/remark-plugin-anchor-link-data/guarantee-unique-slug.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import slugify from 'slugify'
+
+/**
+ * Given a string to be made into a slug,
+ * and given an array of existing slugs with which our new slug cannot collide,
+ *
+ * Return a slug version of the provided string which is guaranteed
+ * to not collide with any strings in the existing slugs array.
+ */
+function guaranteeUniqueSlug(string: string, existingSlugs: string[]): string {
```

Comment:
Lil util used for `anchor-links`, similar to https://github.com/hashicorp/remark-plugins/blob/main/generate_slug.js, but different in a couple ways.

First, it better handles edge cases where there are duplicate-like slugs occur in content. As a contrived example, imagine a document contains first the heading `## Step`, then a duplicate heading `## Step`, and finally another heading `## Step 1`. The previous `generate_slug` would fail to avoid conflicts here, generating the slugs `step`, `step-1`, and `step-1`, with a conflict between the latter slugs.  The revised `guaranteeUniqueSlug` generates the slugs `step`, `step-1`, and `step-1-1`.

Second, unlike `generate_slug`, it does not mutate the `links` argument. Instead, it takes an array of `existingSlugs`. This is intended to allow us to more easily decouple the heading anchor link generation from the "code list item" anchor link generation, as those plugins will no longer have to be in the same execution scope.